src/python/espressomd/electrostatics.pyx

```diff
@@ -200,74 +200,7 @@ IF P3M == 1:
-            Electrostatics prefactor (see :eq:`coulomb_prefactor`).
-        accuracy : :obj:`float`
-            P3M tunes its parameters to provide this target accuracy.
-        alpha : :obj:`float`, optional
-            The Ewald parameter.
-        cao : :obj:`float`, optional
-            The charge-assignment order, an integer between 0 and 7.
-        epsilon : :obj:`float` or :obj:`str`, optional
-            A positive number for the dielectric constant of the
-            surrounding medium. Use ``'metallic'`` to set the dielectric
-            constant of the surrounding medium to infinity (default).
-        mesh : :obj:`int` or (3,) array_like of :obj:`int`, optional
-            The number of mesh points in x, y and z direction. Use a single
-            value for cubic boxes.
-        r_cut : :obj:`float`, optional
-            The real space cutoff.
-        tune : :obj:`bool`, optional
-            Used to activate/deactivate the tuning method on activation.
-            Defaults to ``True``.
-        check_neutrality : :obj:`bool`, optional
-            Raise a warning if the system is not electrically neutral when
-            set to ``True`` (default).
-
-        """
-
-        def __init__(self, *args, **kwargs):
-            super().__init__(*args, **kwargs)
-
-        def validate_params(self):
-            default_params = self.default_params()
-            if not (self._params["prefactor"] > 0.0):
-                raise ValueError("prefactor should be a positive float")
-
-            if is_valid_type(self._params["mesh"], int):
-                if self._params["mesh"] % 2 != 0 and self._params["mesh"] != -1:
-                    raise ValueError(
-                        "P3M requires an even number of mesh points in all directions")
-            else:
-                check_type_or_throw_except(self._params["mesh"], 3, int,
-                                           "P3M mesh has to be an integer or integer list of length 3")
-                if (self._params["mesh"][0] % 2 != 0 and self._params["mesh"][0] != -1) or \
-                   (self._params["mesh"][1] % 2 != 0 and self._params["mesh"][1] != -1) or \
-                   (self._params["mesh"][2] % 2 != 0 and self._params["mesh"][2] != -1):
-                    raise ValueError(
-                        "P3M requires an even number of mesh points in all directions")
-
-            if self._params["epsilon"] == "metallic":
-                self._params["epsilon"] = 0.0
-
-            check_type_or_throw_except(
-                self._params["epsilon"], 1, float,
-                "epsilon should be a double or 'metallic'")
-
-            if self._params["mesh_off"] != default_params["mesh_off"]:
-                check_type_or_throw_except(self._params["mesh_off"], 3, float,
-                                           "mesh_off should be a (3,) array_like of values between 0.0 and 1.0")
-
+    cdef class P3MBase(ElectrostaticInteraction):
```

Comment:
the utility function `_check_and_copy_mesh_size` can be moved into this base class